### ISSUE TITLE
[Bug] 403 when downloading images from danbooru

### ISSUE THREAD

**Issue by narugo1992** [deepghs/waifuc#22]
## How to reproduce

```python
import os.path
from pprint import pprint

import matplotlib.pyplot as plt
import requests
# from curl_cffi import requests  # when use this line, it's okay
from hbutils.system import TemporaryDirectory
from imgutils.data import load_image

from waifuc.config.meta import __TITLE__, __VERSION__
from waifuc.utils import get_requests_session

session = get_requests_session(session=requests.Session())

session.headers.update({
    "User-Agent": f"{__TITLE__}/{__VERSION__}",
})

resp = session.get(
    'https://cdn.donmai.us/original/9b/25/__akisato_konoha_uehara_meiko_shimoda_kaori_yamada_touya_rokuta_mamoru_and_4_more_comic_party_and_1_more__9b257058ee0866d554d01e9036ecb3b6.jpg',
)
print(resp.content)
print(len(resp.content))
pprint(resp.request.headers)
print(resp)

with TemporaryDirectory() as td:
    img_file = os.path.join(td, 'img.png')
    with open(img_file, 'wb') as f:
        f.write(resp.content)

    img = load_image(img_file)
    plt.imshow(img)
    plt.show()

```

**Comment by nonamethanks** [comment]
FYI cloudflare protection was turned on specifically because of bots like this one using up all of our bandwidth and doing [millions of requests without any rate limiting](https://discord.com/channels/310432830138089472/710192137878241372/1190461734595608576). This bot specifically  made 1.92 million requests even after the cloudflare block was put up. 

Kindly implement some kind of rate limiting, or far more aggressive blocks will be used next time.

**Comment by nonamethanks** [comment]
It is a bot because it automatically performs automated functions, namely looping through and downloading potentially thousands of images all at once. And you can very much enforce rate limiting internally. 

Your own README shows how to download 10 images from danbooru. What do you think happens when 100 users decide at the same time that they want to get post searches containing a thousand or more pictures, without any rate limiting? 

Nowhere in your code are there proper waits or backoffs. For example looking at [this function](https://github.com/deepghs/waifuc/blob/6ffed73d259509a49e057f5e700b551c2d8347d9/waifuc/source/danbooru.py#L54), which loops through pages to fetch all posts, there's no sleep anywhere. You only sleep [here](https://github.com/deepghs/waifuc/blob/6ffed73d259509a49e057f5e700b551c2d8347d9/waifuc/utils/session.py#L53) in case of exceptions, but that's for the posts endpoint. If I'm reading the code correctly, when downloading files [you don't even sleep for errors](https://github.com/deepghs/waifuc/blob/6ffed73d259509a49e057f5e700b551c2d8347d9/waifuc/source/web.py#L28). 

Look into libraries like [Pyrate-Limiter](https://github.com/vutran1710/PyrateLimiter) or put some custom backoff, but again, if this persists you leave us no choice but to block this kind of project entirely.

**Comment by narugo1992** [comment]
fixed in #28
